lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+	/* istanbul ignore next */
+	return function () {
+		throw new Error('VIRTUAL METHOD');
+	};
+}
+function notImplementedMethod() {
+	/* istanbul ignore next */
+	return function () {
+		throw new Error('NOT IMPLEMENTED METHOD');
+	};
+}
+
+
+// Structures to define HTTP requests.
+// Support GET and POST.
+// The syntax for creating an instance of a GET request is GET is `GET(url)`.
+// The syntax for creating an instance of a POST request is POST is `POST(url, data, content_type)`'.
+// Alternatively, `POSTFromFile(url, file_name, content_type)` is supported.
+// Both GET and two forms of POST allow `.UserAgent(custom_user_agent)`.
+function HTTPRequestBase(url) {
+	cppArguments.assert('HTTPRequestBase', [
+		[ cppArguments.assertion('string', 'const std::string&', 'url') ]
+	], arguments);
+	
+	this.url = url;
+	this.custom_user_agent = "";
+	this.allow_redirects = false;
+}
+HTTPRequestBase.prototype.UserAgent = function (new_custom_user_agent) {
+	cppArguments.assert('HTTPRequestBase::UserAgent', [
+		[ cppArguments.assertion('string', 'const std::string&', 'new_custom_user_agent') ]
+	], arguments);
+	
+	this.custom_user_agent = new_custom_user_agent;
+	
+	return this;
+};
+HTTPRequestBase.prototype.AllowRedirects = function (allow_redirects_setting) {
+	cppArguments.assert('HTTPRequestBase::AllowRedirects', [
+		[ cppArguments.assertion('bool', 'bool', 'allow_redirects_setting', cppArguments.ASSERTION_MODE_OPTIONAL) ]
+	], arguments);
+	
+	if (typeof allow_redirects_setting === 'undefined') {
+		allow_redirects_setting = true;
+	}
+	
+	this.allow_redirects = allow_redirects_setting;
+	
+	return this;
+};
+
+
+function GET(url) {
+	var _this = this;
+	
+	if (!(_this instanceof GET)) {
+		_this = new GET(url);
+		return _this;
+	}
+	
+	cppArguments.assert('GET', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			function (url) {
+				HTTPRequestBase.call(_this, url);
+			}
+		]
+	], arguments);
+}
+inherits(GET, HTTPRequestBase);
+
+
+function POST(url) {
+	var _this = this;
+	
+	if (!(_this instanceof POST)) {
+		// HACK: Cannot use `apply` to call a constructor with the exact number of arguments.
+		switch (arguments.length) {
+			case 0:
+				// HACK: We know `POST` `cppArguments.assert` will throw on this.
+				_this = new POST();
+				/* istanbul ignore next: previous statement should throw */ break;
+		
+			case 1:
+				_this = new POST(arguments[0]);
+				break;
+		
+			case 2:
+				_this = new POST(arguments[0], arguments[1]);
+				break;
+		
+			case 3:
+				_this = new POST(arguments[0], arguments[1], arguments[2]);
+				break;
+		
+			default:
+				// HACK: We know `POST` `cppArguments.assert` will throw on this.
+				_this = new POST(arguments[0], arguments[1], arguments[2], arguments[3]);
+				/* istanbul ignore next: previous statement should throw */ break;
+		}
+		return _this;
+	}
+	
+	_this.has_body = false;
+	_this.body = "";
+	_this.content_type = "";
+	
+	cppArguments.assert('POST', [
```

Comment:
Since you are passing in functions with logic as well (yay!), perhaps some `cppArguments.dispatch()` (or even `emulateCppFunctionCall()`) is a better name than `cppArguments.assert()`?